src/lib/services/access-service.ts

```diff
@@ -300,7 +381,82 @@ export class AccessService {
     }
 
     async getRootRole(roleName: RoleName): Promise<IRole> {
-        const roles = await this.store.getRootRoles();
+        const roles = await this.roleStore.getRootRoles();
         return roles.find((r) => r.name === roleName);
     }
+
+    async getAllRoles(): Promise<ICustomRole[]> {
+        return this.roleStore.getAll();
+    }
+
+    async createRole(role: IRoleCreation): Promise<ICustomRole> {
+        await this.validateRole(role);
+        const baseRole = {
+            name: role.name,
+            description: role.description,
+            roleType: 'custom',
+        };
+        const rolePermissions = role.permissions;
+        const newRole = await this.roleStore.create(baseRole);
+        if (rolePermissions) {
+            this.store.addEnvironmentPermissionsToRole(
+                newRole.id,
+                rolePermissions,
+            );
+        }
+        return newRole;
+    }
+
+    async updateRole(role: IRoleUpdate): Promise<ICustomRole> {
+        await this.validateRole(role, role.id);
+        const baseRole = {
+            id: role.id,
+            name: role.name,
+            description: role.description,
+            roleType: 'custom',
+        };
+        const rolePermissions = role.permissions;
+        const newRole = await this.roleStore.update(baseRole);
+        if (rolePermissions) {
+            this.store.wipePermissionsFromRole(newRole.id);
```

Comment:
Is it needed to wipe the permissions from the role first? We have a total of three db transactions for this operation, it would be nice if we could reduce it to 2.

Reply:
I really want to say yes but I don't see an obvious way to do so without 3 queries. At the moment, we're running a query per transaction but I really think we should take a look at doing something like this: https://stackoverflow.com/questions/40580674/get-knex-js-transactions-working-with-es7-async-await so that we can get our db mutations atomic.

That being said, if there's something obvious that I missed that would take this down to 2 queries I am more than open!